cmd/kube-scheduler/app/server.go

```diff
@@ -194,25 +194,33 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 		}
 	}
 
-	// Start all informers.
-	cc.InformerFactory.Start(ctx.Done())
-	// DynInformerFactory can be nil in tests.
-	if cc.DynInformerFactory != nil {
-		cc.DynInformerFactory.Start(ctx.Done())
-	}
+	startInformersAndWaitForSync := func(ctx context.Context) {
+		// Start all informers.
+		cc.InformerFactory.Start(ctx.Done())
+		// DynInformerFactory can be nil in tests.
+		if cc.DynInformerFactory != nil {
+			cc.DynInformerFactory.Start(ctx.Done())
+		}
 
-	// Wait for all caches to sync before scheduling.
-	cc.InformerFactory.WaitForCacheSync(ctx.Done())
-	// DynInformerFactory can be nil in tests.
-	if cc.DynInformerFactory != nil {
-		cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
+		// Wait for all caches to sync before scheduling.
+		cc.InformerFactory.WaitForCacheSync(ctx.Done())
+		// DynInformerFactory can be nil in tests.
+		if cc.DynInformerFactory != nil {
+			cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
+		}
+	}
+	if cc.ComponentConfig.MaintainCacheWhenNonLeader || cc.LeaderElection == nil {
+		startInformersAndWaitForSync(ctx)
 	}
 
 	// If leader election is enabled, runCommand via LeaderElector until done and exit.
 	if cc.LeaderElection != nil {
 		cc.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
 			OnStartedLeading: func(ctx context.Context) {
 				close(waitingForLeader)
+				if !cc.ComponentConfig.MaintainCacheWhenNonLeader {
+					startInformersAndWaitForSync(ctx)
```

Comment:
is there any metric or observable signal (/healthz, etc) that would tell us if we got stuck here, holding leader lease and failing to sync / run?

Reply:
Would a log such as "Waiting informer cache to sync..." be sufficient?

Reply:
No. That would require some externally configured log-based metrics to properly mark the pod as unhealthy.

Reply:
I don't know the best way to address this but I add two logs around `startInformersAndWaitForSync(ctx)`. At least, this makes it obvious if the scheduler is stuck here.

Reply:
above, we're setting up `/healthz` endpoints for scheduler in newHealthzAndMetricsHandler ... for kube-apiserver, we also serve /readyz (added in https://github.com/kubernetes/kubernetes/pull/78458) that returns false until informers sync (added in https://github.com/kubernetes/kubernetes/pull/93670)... should we do something similar here?

cc @alculquicondor for scheduler input

cc @logicalhan and @wojtek-t for input on readyz existence and behavior

Reply:
I'm not sure if the "healthy until leader, then unhealthy until informers sync" matches the readyz or healthz pattern better (or if we need a concept of how long we'd tolerate a scheduler that was leader with unsynced informers, similar to the `--livez-grace-period` option kube-apiserver has)

Reply:
> above, we're setting up `/healthz` endpoints for scheduler in newHealthzAndMetricsHandler ... for kube-apiserver, we also serve /readyz (added in #78458) that returns false until informers sync (added in #93670)... should we do something similar here?

Are you suggesting that scheduler add separate `/livez` and `/readyz` endpoints, with the `startInformersAndWaitForSync` being a readiness check? If so, then yes I think agree with that. Throwing `startInformersAndWaitForSync` into the liveness check could be dangerous since it could cause crashlooping behaviors if syncs for resources with large numbers of resources takes too long.

Reply:
Given we're touching this logic now, I would definitely like to introduce better health-related metrics. I think that pretty much only kube-apiserver is exposing something reasonable and we should extend our coverage here.

I agree that splitting into livez/ and readyz/ is what we need here:
- livez/ inherits the existing behavior of healthz/ [it just becomes true once the component starts]
- readyz/ is failing when the component isn't initialized (i.e. it's only true after informers are already synced)

Now what we get is:
- non-leader with delayed caching will always be non-ready
- non-leader without delayed caching will be ready as soon as its cache initializes
- leader currently initializing can be distinguished from non-leader using the leader-election metrics

I think that makes it very clear and I don't personally see a problem with readyz/ being false for extended period of time for non-leaders. @liggitt - do you have any concerns here?

Reply:
Splitting /healthz is a reasonable extension of this PR. I created a separate [issue](https://github.com/kubernetes/kubernetes/issues/118011) to track that work where we can continue the discussion and implementation.

@liggitt would you mind approving this PR if no further changes needed except for this?